arangod/Futures/SharedState.h

```diff
@@ -0,0 +1,292 @@
+  ///   returned reference after it attaches a callback unless it knows that
+  ///   the callback does not move-out the referenced result.)
+  /// - Done: Safe but sometimes unusable. (Always returns a valid reference,
+  ///   but the referenced result may or may not have been modified, including
+  ///   possibly moved-out, depending on what the callback did; some but not
+  ///   all callbacks modify (possibly move-out) the result.)
+  Try<T>& getTry() {
+    TRI_ASSERT(hasResult());
+    return _result;
+  }
+  Try<T> const& getTry() const {
+    TRI_ASSERT(hasResult());
+    return _result;
+  }
+  
+  /// Call only from consumer thread.
+  /// Call only once - else undefined behavior.
+  ///
+  /// See FSM graph for allowed transitions.
+  ///
+  /// If it transitions to Done, synchronously initiates a call to the callback,
+  /// and might also synchronously execute that callback (e.g., if there is no
+  /// executor or if the executor is inline).
+  template <typename F>
+  void setCallback(F&& func) {
+    TRI_ASSERT(!hasCallback());
+    
+    // construct _callback first; if that fails, context_ will not leak
+    _callback = std::forward<F>(func);
+    //::new (&_callback) Callback(std::forward<F>(func));
+    
+    auto state = _state.load(std::memory_order_acquire);
+    while (true) {
+      switch (state) {
+        case State::Start:
+          if (_state.compare_exchange_strong(state, State::OnlyCallback, std::memory_order_release)) {
+            return;
+          }
+          TRI_ASSERT(state == State::OnlyResult); // race with setResult
+          [[gnu::fallthrough]];
+          
+        case State::OnlyResult:
+          if (_state.compare_exchange_strong(state, State::Done, std::memory_order_release)) {
+            doCallback();
+            return;
+          }
+          [[gnu::fallthrough]];
+          
+        default:
+          TRI_ASSERT(false); // unexpected state
+      }
+    }
+  }
+  
+  /// Call only from producer thread.
+  /// Call only once - else undefined behavior.
+  ///
+  /// See FSM graph for allowed transitions.
+  ///
+  /// If it transitions to Done, synchronously initiates a call to the callback,
+  /// and might also synchronously execute that callback (e.g., if there is no
+  /// executor or if the executor is inline).
+  void setResult(Try<T>&& t) {
+    TRI_ASSERT(!hasResult());
+    // call move constructor of content
+    ::new (&_result) Try<T>(std::move(t));
+    
+    auto state = _state.load(std::memory_order_acquire);
+    while (true) {
+      switch (state) {
+        case State::Start:
+          if (_state.compare_exchange_strong(state, State::OnlyResult, std::memory_order_release)) {
+            return;
+          }
+          TRI_ASSERT(state == State::OnlyCallback); // race with setCallback
+          [[gnu::fallthrough]];
+          
+        case State::OnlyCallback:
+          if (_state.compare_exchange_strong(state, State::Done, std::memory_order_release)) {
```

Comment:
This CAS has to use memory_order_acquire to synchronize-with the release-CAS in line 166.